### PR TITLE
Make AuthorityId generic

### PR DESCRIPTION
**Pull request by sorpaas**
Our current implementation pin authority Id to be `primitives::AuthorityId`, which is really specific to ed25519. This makes `AuthorityId` generic by fetching the type from `Block` trait, thus makes it possible for chains to write custom consensus and use other key-signing schemes.

For this PR, I left aura and grandpa to pin ed25519 for now.

### CODE REVIEW

File: core/consensus/common/src/lib.rs [hunk 1]
```diff
@@ -66,9 +65,8 @@ pub trait Environment<B: Block, ConsensusData> {
 	type Error: From<Error>;
 
 	/// Initialize the proposal logic on top of a specific header. Provide
-	/// the authorities at that header, and a local key to sign any additional
-	/// consensus messages with as well.
-	fn init(&self, parent_header: &B::Header, authorities: &[AuthorityId], sign_with: Arc<ed25519::Pair>)
```
Comment (sorpaas) [inline core/consensus/common/src/lib.rs]: It doesn't look like we're using the `sign_with` parameter anywhere and it's specific to ed25519, so I just removed it.

**Comment by gavofyork** [comment]
Please ensure there's a PR waiting for substrate-node-template that keeps it up to date with these API changes.

**Comment by rphmeier** [comment]
I would rather not have `AuthorityId` be part of `Block` -- what about systems where we don't have an `AuthorityId`?

**Comment by sorpaas** [comment]
@rphmeier I agree. We should definitely move authorities out of core and into consensus layers. But this PR is for short-term before we do that -- right now there's already definition of `AuthorityId` within `Block` trait, via `Block::Header::Digest::DigestItem::AuthorityId`. We just didn't utilize that information.

File: core/consensus/common/src/lib.rs (hunk 1 above)
Comment (gnunicorn) [reply core/consensus/common/src/lib.rs]: It has its history in the rhd/bft interface we had before. It is currently not-working and won't be looking the same anyways. We can reintroduce that if we need to (though I doubt this pattern was good in the first place)

**Comment by gnunicorn** [comment]
> I agree. We should definitely move authorities out of core and into consensus layers. But this PR is for short-term before we do that -- right now there's already definition of `AuthorityId` within `Block` trait, via `Block::Header::Digest::DigestItem::AuthorityId`. We just didn't utilize that information.

Could you put this into a ticket and elaborate a bit, @sorpaas ? This sounds like a good starter tickets for people to pick up!

**Comment by sorpaas** [comment]
@gnunicorn I created https://github.com/paritytech/substrate/issues/1365. I don't think that's a starter issue, though. :)

**Pull request merged**